File: Third Year/Semester 5/Parallel and Distributed Programming/Laboratory 1/Example/CPP/mutex.cpp
#include <iostream>
#include <thread>
#include <mutex>

std::mutex mtx;

void critical_section(int id) {
    std::lock_guard<std::mutex> lock(mtx);
    std::cout << "Thread-" << id << " entering critical section" << std::endl;
    std::cout << "Thread-" << id << " leaving critical section" << std::endl;
}

int main() {
    std::thread t1(critical_section, 1);
    std::thread t2(critical_section, 2);
    t1.join();
    t2.join();
    return 0;
}

File: Third Year/Semester 5/Parallel and Distributed Programming/Laboratory 1/Example/Python/mutex.py
import threading

mutex = threading.Lock()

def critical_section(id):
    with mutex:
        print(f"Thread-{id} entering critical section")
        print(f"Thread-{id} leaving critical section")

threads = []
for i in range(5):
    t = threading.Thread(target=critical_section, args=(i,))
    threads.append(t)
    t.start()

for t in threads:
    t.join()